src/Microsoft.ComponentDetection.Detectors/rust/RustCrateDetector.cs

```diff
@@ -34,61 +38,200 @@ protected override Task OnFileFound(ProcessRequest processRequest, IDictionary<s
+                        {
+                            if (packageList.Any(p => p.package.Equals(cargoPackage)))
+                            {
+                                // Ignore duplicate packages
+                                continue;
+                            }
+                        }
+                        else
+                        {
+                            packageList = new List<(CargoPackage, CargoComponent)>();
+                            packagesByName.Add(cargoPackage.name, packageList);
+                        }
+
+                        // Create a node for each non-local package to allow adding dependencies later.
+                        CargoComponent cargoComponent = null;
+                        if (!IsLocalPackage(cargoPackage))
+                        {
+                            cargoComponent = new CargoComponent(cargoPackage.name, cargoPackage.version);
+                            singleFileComponentRecorder.RegisterUsage(new DetectedComponent(cargoComponent));
+                        }
+
+                        // Add the package/component pair to the list
+                        packageList.Add((cargoPackage, cargoComponent));
+                    }
+
+                    // Pass 2: Register dependencies.
+                    foreach (var packageList in packagesByName.Values)
+                    {
+                        // Get the parent package and component
+                        foreach (var (parentPackage, parentComponent) in packageList)
+                        {
+                            if (parentPackage.dependencies == null)
+                            {
+                                // This package has no dependency edges to contribute.
+                                continue;
+                            }
+
+                            // Process each dependency
+                            foreach (var dependency in parentPackage.dependencies)
+                            {
+                                ProcessDependency(cargoLockFile, singleFileComponentRecorder, seenAsDependency, packagesByName, parentPackage, parentComponent, dependency);
+                            }
+                        }
+                    }
+
+                    // Pass 3: Conservatively mark packages we found no dependency to as roots
+                    foreach (var packageList in packagesByName.Values)
+                    {
+                        // Get the package and component.
+                        foreach (var (package, component) in packageList)
+                        {
+                            if (!IsLocalPackage(package) && !seenAsDependency.Contains(package))
+                            {
+                                var detectedComponent = new DetectedComponent(component);
+                                singleFileComponentRecorder.RegisterUsage(detectedComponent, isExplicitReferencedDependency: true);
```

Comment:
What happened to dev dependency detection? Is that being considered in these updates, if so, how will those be treated now? Users won't like to start picking up dependencies as regular build time dependencies when they used to be dev dependencies.

Reply:
This does not report dev dependencies separately (e.g., they are included alongside other dependencies as well) as that information is not currently included in the lock file. Unfortunately, there isn't an obvious middle ground here between having something that is simple and robust (lock file processing) or replicating the full functionality of cargo.

This change puts us in the position of not missing dependencies, with the option to (in the future) introduce a tool that uses cargo to get a more accurate picture (either directly or by consuming SBOMs generated from cargo).